File: services/backend/app/data/telemetry/model/telemetry.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel

# This schema is duplicated at services/companion/app/models.py.
# If you change it here, change it there.


class Position(BaseModel):
    lat: float | None
    lon: float | None
    rel_alt: float | None
    abs_alt: float | None


class Gps(BaseModel):
    num_satellites: int | None
    fix_type: str | None


class Health(BaseModel):
    is_gyrometer_calibrated: bool | None
    is_accelerometer_calibrated: bool | None
    is_magnetometer_calibrated: bool | None
    is_local_position_ok: bool | None
    is_global_position_ok: bool | None
    is_home_position_ok: bool | None
    is_armable: bool | None


class MavlinkTelemetry(BaseModel):
    flight_controller_last_seen: datetime
    is_armed: bool
    is_in_air: bool
    flight_mode: str
    battery_percentage: float | None
    flight_time_remaining: float | None
    position: Position | None
    gps: Gps | None
    health: Health | None


class CompanionState(str, Enum):
    CONNECTING = "connecting"
    READY = "ready"


class Telemetry(BaseModel):
    companion_state: CompanionState
    companion_state_timestamp: datetime
    mavlink_telemetry: MavlinkTelemetry | None

File: services/ui/src/types/telemetry.ts
/**
 * Hand-written mirror of the backend telemetry schema
 * (services/backend/app/data/telemetry/model/telemetry.py).
 * If you change it there, change it here.
 */

export type CompanionState = "connecting" | "ready"

export interface Position {
  lat: number | null
  lon: number | null
  rel_alt: number | null
  abs_alt: number | null
}

export interface Gps {
  num_satellites: number | null
  fix_type: string | null
}

export interface Health {
  is_gyrometer_calibrated: boolean | null
  is_accelerometer_calibrated: boolean | null
  is_magnetometer_calibrated: boolean | null
  is_local_position_ok: boolean | null
  is_global_position_ok: boolean | null
  is_home_position_ok: boolean | null
  is_armable: boolean | null
}

export interface MavlinkTelemetry {
  /** ISO 8601 datetime, companion clock */
  flight_controller_last_seen: string
  is_armed: boolean
  is_in_air: boolean
  flight_mode: string
  battery_percentage: number | null
  flight_time_remaining: number | null
  position: Position | null
  gps: Gps | null
  health: Health | null
}

export interface Telemetry {
  companion_state: CompanionState
  /** ISO 8601 datetime, companion clock */
  companion_state_timestamp: string
  mavlink_telemetry: MavlinkTelemetry | null
}
